In [47]:
import pandas as pd 
import matplotlib.pyplot as plt

In [48]:
approval_date_1990_1999_path = r'/Users/user/Documents/htx_xdata/htx_xdata_tech_interview/data/HDB/resale-flat-prices-based-on-approval-date-1990-1999.csv'
approval_date_2000_feb_2012_path = r'/Users/user/Documents/htx_xdata/htx_xdata_tech_interview/data/HDB/resale-flat-prices-based-on-approval-date-2000-feb-2012.csv'

registration_date_mar_2012_dec_2014_path = r'/Users/user/Documents/htx_xdata/htx_xdata_tech_interview/data/HDB/resale-flat-prices-based-on-registration-date-from-mar-2012-to-dec-2014.csv'
registration_date_jan_2015_dec_2016_path = r'/Users/user/Documents/htx_xdata/htx_xdata_tech_interview/data/HDB/resale-flat-prices-based-on-registration-date-from-jan-2015-to-dec-2016.csv'
registration_date_jan_2017_path = r'/Users/user/Documents/htx_xdata/htx_xdata_tech_interview/data/HDB/resale-flat-prices-based-on-registration-date-from-jan-2017-onwards.csv'

In [49]:
approval_date_1990_1999_df = pd.read_csv(approval_date_1990_1999_path)
approval_date_2000_feb_2012_df = pd.read_csv(approval_date_2000_feb_2012_path)

registration_date_mar_2012_dec_2014_df = pd.read_csv(registration_date_mar_2012_dec_2014_path)
registration_date_jan_2015_dec_2016_df = pd.read_csv(registration_date_jan_2015_dec_2016_path)
registration_date_jan_2017_df = pd.read_csv(registration_date_jan_2017_path)

Exploratory Data Analysis
1. Structural Investigation
Check for

- predictor and target
- Missing values
- Duplicated values
- Type of variable, Data type and Variable Category
- Categorical Variable Cardinality

In [50]:
approval_date_1990_1999_df

month        town  flat_type block       street_name storey_range  \
0       1990-01  ANG MO KIO     1 ROOM   309  ANG MO KIO AVE 1     10 TO 12   
1       1990-01  ANG MO KIO     1 ROOM   309  ANG MO KIO AVE 1     04 TO 06   
2       1990-01  ANG MO KIO     1 ROOM   309  ANG MO KIO AVE 1     10 TO 12   
3       1990-01  ANG MO KIO     1 ROOM   309  ANG MO KIO AVE 1     07 TO 09   
4       1990-01  ANG MO KIO     3 ROOM   216  ANG MO KIO AVE 1     04 TO 06   
...         ...         ...        ...   ...               ...          ...   
287195  1999-12      YISHUN  EXECUTIVE   611      YISHUN ST 61     10 TO 12   
287196  1999-12      YISHUN  EXECUTIVE   324       YISHUN CTRL     01 TO 03   
287197  1999-12      YISHUN  EXECUTIVE   392      YISHUN AVE 6     07 TO 09   
287198  1999-12      YISHUN  EXECUTIVE   356    YISHUN RING RD     04 TO 06   
287199  1999-12      YISHUN  EXECUTIVE   358    YISHUN RING RD     01 TO 03   

        floor_area_sqm      flat_model  lease_commence_date  resale_price  
0                 31.0        IMPROVED                 1977          9000  
1                 31.0        IMPROVED                 1977          6000  
2                 31.0        IMPROVED                 1977          8000  
3                 31.0        IMPROVED                 1977          6000  
4                 73.0  NEW GENERATION                 1976         47200  
...                ...             ...                  ...           ...  
287195           142.0       APARTMENT                 1987        456000  
287196           142.0       APARTMENT                 1988        408000  
287197           146.0      MAISONETTE                 1988        469000  
287198           146.0      MAISONETTE                 1988        440000  
287199           145.0      MAISONETTE                 1988        484000  

[287200 rows x 10 columns]

In [51]:
# Check the shape of the dataframe
print(approval_date_1990_1999_df.shape)
print('---------------------------------------')
# Check the datatype information of the features
print(approval_date_1990_1999_df.info())
print('---------------------------------------')
# Check for missing values in each column
print(approval_date_1990_1999_df.isnull().sum())
print('---------------------------------------')
# check for missing values in each column as a percentage of whole dataset
print((approval_date_1990_1999_df.isnull().sum() / approval_date_1990_1999_df.shape[0]) * 100)
print('---------------------------------------')
# Check for duplicates in the dataframe
print(approval_date_1990_1999_df.duplicated().sum())
print('---------------------------------------')

(287200, 10)
---------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 287200 entries, 0 to 287199
Data columns (total 10 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   month                287200 non-null  object 
 1   town                 287200 non-null  object 
 2   flat_type            287200 non-null  object 
 3   block                287200 non-null  object 
 4   street_name          287200 non-null  object 
 5   storey_range         287200 non-null  object 
 6   floor_area_sqm       287200 non-null  float64
 7   flat_model           287200 non-null  object 
 8   lease_commence_date  287200 non-null  int64  
 9   resale_price         287200 non-null  int64  
dtypes: float64(1), int64(2), object(7)
memory usage: 21.9+ MB
None
---------------------------------------
month                  0
town                   0
flat_type              0
block                  0
street_name     

In [52]:
# approval_date_1990_1999_duplicated_df = approval_date_1990_1999_df[approval_date_1990_1999_df.duplicated()]
# print(approval_date_1990_1999_duplicated_df.shape)
# approval_date_1990_1999_duplicated_df[(approval_date_1990_1999_duplicated_df['town']=='GEYLANG') & (approval_date_1990_1999_duplicated_df['flat_model']=='STANDARD')]

# Show all rows that are duplicates
approval_date_1990_1999_df_duplicates = approval_date_1990_1999_df[approval_date_1990_1999_df.duplicated(keep=False)]
print("Approval date 1990 1999 duplicate rows:")
approval_date_1990_1999_df_duplicates.head(20)

Approval date 1990 1999 duplicate rows:


month         town  flat_type block        street_name storey_range  \
672   1990-01      GEYLANG     3 ROOM    47         CIRCUIT RD     01 TO 03   
673   1990-01      GEYLANG     3 ROOM    47         CIRCUIT RD     01 TO 03   
725   1990-01      HOUGANG     3 ROOM   308      HOUGANG AVE 5     10 TO 12   
726   1990-01      HOUGANG     3 ROOM   308      HOUGANG AVE 5     10 TO 12   
842   1990-01  JURONG WEST     3 ROOM   145        HU CHING RD     04 TO 06   
843   1990-01  JURONG WEST     3 ROOM   145        HU CHING RD     04 TO 06   
851   1990-01  JURONG WEST     3 ROOM   172        BOON LAY DR     10 TO 12   
854   1990-01  JURONG WEST     3 ROOM   172        BOON LAY DR     10 TO 12   
856   1990-01  JURONG WEST     3 ROOM   185       BOON LAY AVE     22 TO 24   
862   1990-01  JURONG WEST     3 ROOM   185       BOON LAY AVE     22 TO 24   
863   1990-01  JURONG WEST     3 ROOM   186       BOON LAY AVE     07 TO 09   
867   1990-01  JURONG WEST     3 ROOM   186       BOON LAY AVE     07 TO 09   
874   1990-01  JURONG WEST     3 ROOM   187       BOON LAY AVE     19 TO 21   
877   1990-01  JURONG WEST     3 ROOM   187       BOON LAY AVE     19 TO 21   
967   1990-01  JURONG WEST  EXECUTIVE   407  JURONG WEST ST 42     10 TO 12   
968   1990-01  JURONG WEST  EXECUTIVE   407  JURONG WEST ST 42     10 TO 12   
1095  1990-01   QUEENSTOWN     3 ROOM     2        GHIM MOH RD     04 TO 06   
1096  1990-01   QUEENSTOWN     3 ROOM     2        GHIM MOH RD     10 TO 12   
1097  1990-01   QUEENSTOWN     3 ROOM     2        GHIM MOH RD     10 TO 12   
1098  1990-01   QUEENSTOWN     3 ROOM     2        GHIM MOH RD     04 TO 06   

      floor_area_sqm      flat_model  lease_commence_date  resale_price  
672             56.0        STANDARD                 1969         18000  
673             56.0        STANDARD                 1969         18000  
725             67.0  NEW GENERATION                 1983         47000  
726             67.0  NEW GENERATION                 1983         47000  
842             64.0        IMPROVED                 1976         23400  
843             64.0        IMPROVED                 1976         23400  
851             70.0        IMPROVED                 1974         32900  
854             70.0        IMPROVED                 1974         32900  
856             59.0        IMPROVED                 1975         22500  
862             59.0        IMPROVED                 1975         22500  
863             59.0        IMPROVED                 1975         31500  
867             59.0        IMPROVED                 1975         31500  
874             68.0        IMPROVED                 1978         26000  
877             68.0        IMPROVED                 1978         26000  
967            150.0      MAISONETTE                 1985        160000  
968            150.0      MAISONETTE                 1985        160000  
1095            65.0        IMPROVED                 1976         36000  
1096            65.0        IMPROVED                 1976         40000  
1097            65.0        IMPROVED                 1976         40000  
1098            65.0        IMPROVED                 1976         36000

### Approval_date_1990_1999_df structural observation
- There are 10 columns and 287,200 rows 
- There are 3 datatypes : int64, float64 and object
- There are 826 duplicated rows 
    - The chances of selling the house at the same price are rare. Hence, we will assume it to be duplicated entries.

In [53]:
approval_date_2000_feb_2012_df

month        town  flat_type block       street_name storey_range  \
0       2000-01  ANG MO KIO     3 ROOM   170  ANG MO KIO AVE 4     07 TO 09   
1       2000-01  ANG MO KIO     3 ROOM   174  ANG MO KIO AVE 4     04 TO 06   
2       2000-01  ANG MO KIO     3 ROOM   216  ANG MO KIO AVE 1     07 TO 09   
3       2000-01  ANG MO KIO     3 ROOM   215  ANG MO KIO AVE 1     07 TO 09   
4       2000-01  ANG MO KIO     3 ROOM   218  ANG MO KIO AVE 1     07 TO 09   
...         ...         ...        ...   ...               ...          ...   
369646  2012-02      YISHUN     5 ROOM   212      YISHUN ST 21     10 TO 12   
369647  2012-02      YISHUN     5 ROOM   758      YISHUN ST 72     01 TO 03   
369648  2012-02      YISHUN     5 ROOM   873      YISHUN ST 81     01 TO 03   
369649  2012-02      YISHUN  EXECUTIVE   664      YISHUN AVE 4     07 TO 09   
369650  2012-02      YISHUN  EXECUTIVE   293      YISHUN ST 22     04 TO 06   

        floor_area_sqm      flat_model  lease_commence_date  resale_price  
0                 69.0        Improved                 1986      147000.0  
1                 61.0        Improved                 1986      144000.0  
2                 73.0  New Generation                 1976      159000.0  
3                 73.0  New Generation                 1976      167000.0  
4                 67.0  New Generation                 1976      163000.0  
...                ...             ...                  ...           ...  
369646           121.0        Improved                 1985      476888.0  
369647           122.0        Improved                 1986      490000.0  
369648           122.0        Improved                 1988      488000.0  
369649           181.0       Apartment                 1992      705000.0  
369650           163.0       Apartment                 1992      630000.0  

[369651 rows x 10 columns]

In [58]:
# Check the shape of the dataframe
print(approval_date_2000_feb_2012_df.shape)
print('---------------------------------------')
# Check the datatype information of the features
print(approval_date_2000_feb_2012_df.info())
print('---------------------------------------')
# Check for missing values in each column
print(approval_date_2000_feb_2012_df.isnull().sum())
print('---------------------------------------')
# check for missing values in each column as a percentage of whole dataset
print((approval_date_2000_feb_2012_df.isnull().sum() / approval_date_2000_feb_2012_df.shape[0]) * 100)
print('---------------------------------------')
# Check for duplicates in the dataframe
print(approval_date_2000_feb_2012_df.duplicated().sum())
print('---------------------------------------')

(369651, 10)
---------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 369651 entries, 0 to 369650
Data columns (total 10 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   month                369651 non-null  object 
 1   town                 369651 non-null  object 
 2   flat_type            369651 non-null  object 
 3   block                369651 non-null  object 
 4   street_name          369651 non-null  object 
 5   storey_range         369651 non-null  object 
 6   floor_area_sqm       369651 non-null  float64
 7   flat_model           369651 non-null  object 
 8   lease_commence_date  369651 non-null  int64  
 9   resale_price         369651 non-null  float64
dtypes: float64(2), int64(1), object(7)
memory usage: 28.2+ MB
None
---------------------------------------
month                  0
town                   0
flat_type              0
block                  0
street_name     

In [60]:
approval_date_2000_feb_2012_df_duplicates = approval_date_2000_feb_2012_df[approval_date_2000_feb_2012_df.duplicated(keep=False)]
print("Approval date 2000 Feb 2012 duplicate rows:")
approval_date_2000_feb_2012_df_duplicates.head(20)

Approval date 2000 Feb 2012 duplicate rows:


month         town  flat_type block         street_name storey_range  \
21    2000-01   ANG MO KIO     3 ROOM   443   ANG MO KIO AVE 10     07 TO 09   
22    2000-01   ANG MO KIO     3 ROOM   443   ANG MO KIO AVE 10     07 TO 09   
175   2000-01        BEDOK     3 ROOM    32       CHAI CHEE AVE     13 TO 15   
176   2000-01        BEDOK     3 ROOM    32       CHAI CHEE AVE     13 TO 15   
284   2000-01        BEDOK     4 ROOM   421        BEDOK NTH RD     13 TO 15   
285   2000-01        BEDOK     4 ROOM   421        BEDOK NTH RD     13 TO 15   
867   2000-01     CLEMENTI     3 ROOM   518       WEST COAST RD     10 TO 12   
868   2000-01     CLEMENTI     3 ROOM   518       WEST COAST RD     10 TO 12   
2326  2000-01       YISHUN     4 ROOM   249        YISHUN AVE 9     10 TO 12   
2327  2000-01       YISHUN     4 ROOM   249        YISHUN AVE 9     10 TO 12   
2475  2000-02   ANG MO KIO     3 ROOM   646    ANG MO KIO AVE 6     07 TO 09   
2476  2000-02   ANG MO KIO     3 ROOM   646    ANG MO KIO AVE 6     07 TO 09   
3386  2000-02      HOUGANG  EXECUTIVE   141          LOR AH SOO     04 TO 06   
3387  2000-02      HOUGANG  EXECUTIVE   141          LOR AH SOO     04 TO 06   
3421  2000-02  JURONG EAST     4 ROOM   219   JURONG EAST ST 21     07 TO 09   
3422  2000-02  JURONG EAST     4 ROOM   219   JURONG EAST ST 21     07 TO 09   
4801  2000-03   ANG MO KIO     3 ROOM   513    ANG MO KIO AVE 8     07 TO 09   
4802  2000-03   ANG MO KIO     3 ROOM   513    ANG MO KIO AVE 8     07 TO 09   
4952  2000-03        BEDOK     3 ROOM   713  BEDOK RESERVOIR RD     04 TO 06   
4953  2000-03        BEDOK     3 ROOM   713  BEDOK RESERVOIR RD     04 TO 06   

      floor_area_sqm      flat_model  lease_commence_date  resale_price  
21              67.0  New Generation                 1979      162000.0  
22              67.0  New Generation                 1979      162000.0  
175             65.0        Improved                 1974      165000.0  
176             65.0        Improved                 1974      165000.0  
284             92.0  New Generation                 1979      275000.0  
285             92.0  New Generation                 1979      275000.0  
867             68.0  New Generation                 1980      154000.0  
868             68.0  New Generation                 1980      154000.0  
2326            84.0      Simplified                 1985      195000.0  
2327            84.0      Simplified                 1985      195000.0  
2475            75.0  New Generation                 1980      188000.0  
2476            75.0  New Generation                 1980      188000.0  
3386           148.0      Maisonette                 1985      448000.0  
3387           148.0      Maisonette                 1985      448000.0  
3421            92.0  New Generation                 1985      248000.0  
3422            92.0  New Generation                 1985      248000.0  
4801            67.0  New Generation                 1984      175000.0  
4802            67.0  New Generation                 1984      175000.0  
4952            68.0  New Generation                 1981      160000.0  
4953            68.0  New Generation                 1981      160000.0

### Approval_date_2000_feb_2012_df structural observation
- There are 10 columns and 369,651rows 
- There are 3 datatypes : int64, float64 and object
- There are 513 duplicated rows 
    - The chances of selling the house at the same price are rare. Hence, we will assume it to be duplicated entries.
- Approval_date_1990_1999_df resale_price is int64 while Approval_date_2000_feb_2012_df is float64

In [55]:
registration_date_mar_2012_dec_2014_df


month        town  flat_type block        street_name storey_range  \
0      2012-03  ANG MO KIO     2 ROOM   172   ANG MO KIO AVE 4     06 TO 10   
1      2012-03  ANG MO KIO     2 ROOM   510   ANG MO KIO AVE 8     01 TO 05   
2      2012-03  ANG MO KIO     3 ROOM   610   ANG MO KIO AVE 4     06 TO 10   
3      2012-03  ANG MO KIO     3 ROOM   474  ANG MO KIO AVE 10     01 TO 05   
4      2012-03  ANG MO KIO     3 ROOM   604   ANG MO KIO AVE 5     06 TO 10   
...        ...         ...        ...   ...                ...          ...   
52198  2014-12      YISHUN     5 ROOM   816       YISHUN ST 81     10 TO 12   
52199  2014-12      YISHUN  EXECUTIVE   325        YISHUN CTRL     10 TO 12   
52200  2014-12      YISHUN  EXECUTIVE   618     YISHUN RING RD     07 TO 09   
52201  2014-12      YISHUN  EXECUTIVE   277       YISHUN ST 22     07 TO 09   
52202  2014-12      YISHUN  EXECUTIVE   277       YISHUN ST 22     04 TO 06   

       floor_area_sqm      flat_model  lease_commence_date  resale_price  
0                45.0        Improved                 1986      250000.0  
1                44.0        Improved                 1980      265000.0  
2                68.0  New Generation                 1980      315000.0  
3                67.0  New Generation                 1984      320000.0  
4                67.0  New Generation                 1980      321000.0  
...               ...             ...                  ...           ...  
52198           122.0        Improved                 1988      580000.0  
52199           146.0      Maisonette                 1988      540000.0  
52200           164.0       Apartment                 1992      738000.0  
52201           152.0      Maisonette                 1985      592000.0  
52202           146.0      Maisonette                 1985      545000.0  

[52203 rows x 10 columns]

In [61]:
# Check the shape of the dataframe
print(registration_date_mar_2012_dec_2014_df.shape)
print('---------------------------------------')
# Check the datatype information of the features
print(registration_date_mar_2012_dec_2014_df.info())
print('---------------------------------------')
# Check for missing values in each column
print(registration_date_mar_2012_dec_2014_df.isnull().sum())
print('---------------------------------------')
# check for missing values in each column as a percentage of whole dataset
print((registration_date_mar_2012_dec_2014_df.isnull().sum() / registration_date_mar_2012_dec_2014_df.shape[0]) * 100)
print('---------------------------------------')
# Check for duplicates in the dataframe
print(registration_date_mar_2012_dec_2014_df.duplicated().sum())
print('---------------------------------------')

(52203, 10)
---------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52203 entries, 0 to 52202
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   month                52203 non-null  object 
 1   town                 52203 non-null  object 
 2   flat_type            52203 non-null  object 
 3   block                52203 non-null  object 
 4   street_name          52203 non-null  object 
 5   storey_range         52203 non-null  object 
 6   floor_area_sqm       52203 non-null  float64
 7   flat_model           52203 non-null  object 
 8   lease_commence_date  52203 non-null  int64  
 9   resale_price         52203 non-null  float64
dtypes: float64(2), int64(1), object(7)
memory usage: 4.0+ MB
None
---------------------------------------
month                  0
town                   0
flat_type              0
block                  0
street_name            0
storey_

In [62]:
registration_date_mar_2012_dec_2014_df_duplicates = registration_date_mar_2012_dec_2014_df[registration_date_mar_2012_dec_2014_df.duplicated(keep=False)]
print("Registration date March 2012 Dec 2014 duplicate rows:")
registration_date_mar_2012_dec_2014_df_duplicates.head(20)

Registration date March 2012 Dec 2014 duplicate rows:


month         town flat_type block          street_name storey_range  \
336    2012-03  BUKIT BATOK    3 ROOM   212       BT BATOK ST 21     06 TO 10   
337    2012-03  BUKIT BATOK    3 ROOM   212       BT BATOK ST 21     06 TO 10   
2650   2012-04  BUKIT BATOK    3 ROOM   167  BT BATOK WEST AVE 8     01 TO 05   
2651   2012-04  BUKIT BATOK    3 ROOM   167  BT BATOK WEST AVE 8     01 TO 05   
3259   2012-04  JURONG EAST    3 ROOM   303    JURONG EAST ST 32     01 TO 05   
3260   2012-04  JURONG EAST    3 ROOM   303    JURONG EAST ST 32     01 TO 05   
4164   2012-04    WOODLANDS    3 ROOM    15       MARSILING LANE     01 TO 05   
4165   2012-04    WOODLANDS    3 ROOM    15       MARSILING LANE     01 TO 05   
4420   2012-04       YISHUN    3 ROOM   740         YISHUN AVE 5     06 TO 10   
4421   2012-04       YISHUN    3 ROOM   740         YISHUN AVE 5     06 TO 10   
5538   2012-05  JURONG WEST    3 ROOM   914    JURONG WEST ST 91     01 TO 05   
5539   2012-05  JURONG WEST    3 ROOM   914    JURONG WEST ST 91     01 TO 05   
8020   2012-06      PUNGGOL    5 ROOM  126D      EDGEDALE PLAINS     01 TO 03   
8021   2012-06      PUNGGOL    5 ROOM  126D      EDGEDALE PLAINS     01 TO 03   
9730   2012-07  JURONG EAST    3 ROOM   231    JURONG EAST ST 21     04 TO 06   
9731   2012-07  JURONG EAST    3 ROOM   231    JURONG EAST ST 21     04 TO 06   
9742   2012-07  JURONG EAST    3 ROOM   110    JURONG EAST ST 13     10 TO 12   
9743   2012-07  JURONG EAST    3 ROOM   110    JURONG EAST ST 13     10 TO 12   
14951  2012-10        BEDOK    3 ROOM    37        CHAI CHEE AVE     04 TO 06   
14952  2012-10        BEDOK    3 ROOM    37        CHAI CHEE AVE     04 TO 06   

       floor_area_sqm      flat_model  lease_commence_date  resale_price  
336              67.0  New Generation                 1985      341000.0  
337              67.0  New Generation                 1985      341000.0  
2650             74.0         Model A                 1984      333000.0  
2651             74.0         Model A                 1984      333000.0  
3259             67.0  New Generation                 1983      330000.0  
3260             67.0  New Generation                 1983      330000.0  
4164             65.0        Improved                 1976      280000.0  
4165             65.0        Improved                 1976      280000.0  
4420             67.0  New Generation                 1985      341000.0  
4421             67.0  New Generation                 1985      341000.0  
5538             73.0         Model A                 1988      312000.0  
5539             73.0         Model A                 1988      312000.0  
8020            110.0        Improved                 2003      510000.0  
8021            110.0        Improved                 2003      510000.0  
9730             67.0  New Generation                 1981      335000.0  
9731             67.0  New Generation                 1981      335000.0  
9742             68.0  New Generation                 1982      360000.0  
9743             68.0  New Generation                 1982      360000.0  
14951            66.0        Improved                 1983      345000.0  
14952            66.0        Improved                 1983      345000.0

### Registration_date_mar_2012_dec_2014_df structural observation
- There are 10 columns and 52,203 rows 
- There are 3 datatypes : int64, float64 and object
- There are 248 duplicated rows 
    - The chances of selling the house at the same price are rare. Hence, we will assume it to be duplicated entries.
- Approval_date_1990_1999_df resale_price is int64 while Approval_date_2000_feb_2012_df is float64
Registration_date_mar_2012_dec_2014_df is float64

In [56]:
registration_date_jan_2015_dec_2016_df


month        town         flat_type block        street_name  \
0      2015-01  ANG MO KIO            3 ROOM   174   ANG MO KIO AVE 4   
1      2015-01  ANG MO KIO            3 ROOM   541  ANG MO KIO AVE 10   
2      2015-01  ANG MO KIO            3 ROOM   163   ANG MO KIO AVE 4   
3      2015-01  ANG MO KIO            3 ROOM   446  ANG MO KIO AVE 10   
4      2015-01  ANG MO KIO            3 ROOM   557  ANG MO KIO AVE 10   
...        ...         ...               ...   ...                ...   
37148  2016-12      YISHUN            5 ROOM   297       YISHUN ST 20   
37149  2016-12      YISHUN            5 ROOM   838       YISHUN ST 81   
37150  2016-12      YISHUN         EXECUTIVE   664       YISHUN AVE 4   
37151  2016-12      YISHUN         EXECUTIVE   325        YISHUN CTRL   
37152  2016-12      YISHUN  MULTI-GENERATION   666       YISHUN AVE 4   

      storey_range  floor_area_sqm        flat_model  lease_commence_date  \
0         07 TO 09            60.0          Improved                 1986   
1         01 TO 03            68.0    New Generation                 1981   
2         01 TO 03            69.0    New Generation                 1980   
3         01 TO 03            68.0    New Generation                 1979   
4         07 TO 09            68.0    New Generation                 1980   
...            ...             ...               ...                  ...   
37148     13 TO 15           112.0          Improved                 2000   
37149     01 TO 03           122.0          Improved                 1987   
37150     10 TO 12           181.0         Apartment                 1992   
37151     01 TO 03           146.0        Maisonette                 1988   
37152     10 TO 12           164.0  Multi Generation                 1987   

       remaining_lease  resale_price  
0                   70      255000.0  
1                   65      275000.0  
2                   64      285000.0  
3                   63      290000.0  
4                   64      290000.0  
...                ...           ...  
37148               82      488000.0  
37149               69      455000.0  
37150               74      778000.0  
37151               70      575000.0  
37152               70      735000.0  

[37153 rows x 11 columns]

In [64]:
# Check the shape of the dataframe
print(registration_date_jan_2015_dec_2016_df.shape)
print('---------------------------------------')
# Check the datatype information of the features
print(registration_date_jan_2015_dec_2016_df.info())
print('---------------------------------------')
# Check for missing values in each column
print(registration_date_jan_2015_dec_2016_df.isnull().sum())
print('---------------------------------------')
# check for missing values in each column as a percentage of whole dataset
print((registration_date_jan_2015_dec_2016_df.isnull().sum() / registration_date_jan_2015_dec_2016_df.shape[0]) * 100)
print('---------------------------------------')
# Check for duplicates in the dataframe
print(registration_date_jan_2015_dec_2016_df.duplicated().sum())
print('---------------------------------------')

(37153, 11)
---------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 37153 entries, 0 to 37152
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   month                37153 non-null  object 
 1   town                 37153 non-null  object 
 2   flat_type            37153 non-null  object 
 3   block                37153 non-null  object 
 4   street_name          37153 non-null  object 
 5   storey_range         37153 non-null  object 
 6   floor_area_sqm       37153 non-null  float64
 7   flat_model           37153 non-null  object 
 8   lease_commence_date  37153 non-null  int64  
 9   remaining_lease      37153 non-null  int64  
 10  resale_price         37153 non-null  float64
dtypes: float64(2), int64(2), object(7)
memory usage: 3.1+ MB
None
---------------------------------------
month                  0
town                   0
flat_type              0
block  

In [65]:
registration_date_jan_2015_dec_2016_df_duplicates = registration_date_jan_2015_dec_2016_df[registration_date_jan_2015_dec_2016_df.duplicated(keep=False)]
print("Registration date Jan 2015 Dec 2016 duplicate rows:")
registration_date_jan_2015_dec_2016_df_duplicates.head(20)

Registration date Jan 2015 Dec 2016 duplicate rows:


month             town flat_type block         street_name  \
660    2015-01  KALLANG/WHAMPOA    3 ROOM    57       GEYLANG BAHRU   
661    2015-01  KALLANG/WHAMPOA    3 ROOM    57       GEYLANG BAHRU   
2165   2015-02         TAMPINES    3 ROOM   403      TAMPINES ST 41   
2166   2015-02         TAMPINES    3 ROOM   403      TAMPINES ST 41   
3895   2015-04            BEDOK    4 ROOM   701  BEDOK RESERVOIR RD   
3896   2015-04            BEDOK    4 ROOM   701  BEDOK RESERVOIR RD   
4245   2015-04         CLEMENTI    3 ROOM   714  CLEMENTI WEST ST 2   
4246   2015-04         CLEMENTI    3 ROOM   714  CLEMENTI WEST ST 2   
5492   2015-05            BEDOK    3 ROOM   409     BEDOK NTH AVE 2   
5493   2015-05            BEDOK    3 ROOM   409     BEDOK NTH AVE 2   
6075   2015-05      JURONG WEST    3 ROOM   501   JURONG WEST ST 51   
6076   2015-05      JURONG WEST    3 ROOM   501   JURONG WEST ST 51   
7090   2015-06            BEDOK    3 ROOM   410     BEDOK NTH AVE 2   
7091   2015-06            BEDOK    3 ROOM   410     BEDOK NTH AVE 2   
7180   2015-06           BISHAN    4 ROOM   267        BISHAN ST 24   
7182   2015-06           BISHAN    4 ROOM   267        BISHAN ST 24   
9113   2015-07    CHOA CHU KANG    5 ROOM  690D  CHOA CHU KANG CRES   
9114   2015-07    CHOA CHU KANG    5 ROOM  690D  CHOA CHU KANG CRES   
10214  2015-08       ANG MO KIO    3 ROOM   473   ANG MO KIO AVE 10   
10215  2015-08       ANG MO KIO    3 ROOM   473   ANG MO KIO AVE 10   

      storey_range  floor_area_sqm      flat_model  lease_commence_date  \
660       16 TO 18            65.0        Improved                 1974   
661       16 TO 18            65.0        Improved                 1974   
2165      07 TO 09            69.0        Improved                 1985   
2166      07 TO 09            69.0        Improved                 1985   
3895      10 TO 12            93.0  New Generation                 1980   
3896      10 TO 12            93.0  New Generation                 1980   
4245      10 TO 12            67.0  New Generation                 1980   
4246      10 TO 12            67.0  New Generation                 1980   
5492      07 TO 09            68.0  New Generation                 1980   
5493      07 TO 09            68.0  New Generation                 1980   
6075      01 TO 03            74.0         Model A                 1986   
6076      01 TO 03            74.0         Model A                 1986   
7090      04 TO 06            68.0  New Generation                 1980   
7091      04 TO 06            68.0  New Generation                 1980   
7180      04 TO 06           104.0         Model A                 1992   
7182      04 TO 06           104.0         Model A                 1992   
9113      13 TO 15           110.0        Improved                 2003   
9114      13 TO 15           110.0        Improved                 2003   
10214     07 TO 09            67.0  New Generation                 1984   
10215     07 TO 09            67.0  New Generation                 1984   

       remaining_lease  resale_price  
660                 58      315000.0  
661                 58      315000.0  
2165                69      350000.0  
2166                69      350000.0  
3895                64      400000.0  
3896                64      400000.0  
4245                64      300000.0  
4246                64      300000.0  
5492                63      340000.0  
5493                63      340000.0  
6075                69      304000.0  
6076                69      304000.0  
7090                63      340000.0  
7091                63      340000.0  
7180                76      530000.0  
7182                76      530000.0  
9113                87      390000.0  
9114                87      390000.0  
10214               67      308000.0  
10215               67      308000.0

### Registration_date_jan_2015_dec_2016_df structural observation
- There are 11 columns and 37,153 rows 
- There are 3 datatypes : int64, float64 and object
- There are 24 duplicated rows 
    - The chances of selling the house at the same price are rare. Hence, we will assume it to be duplicated entries.
- Approval_date_1990_1999_df resale_price is int64 while Approval_date_2000_feb_2012_df is float64,
Registration_date_mar_2012_dec_2014_df is float64,
registration_date_jan_2015_dec_2016_df is float64
- From registration_date_jan_2015_dec_2016_df, there is an extra column callled 'remaining_lease'

In [57]:
registration_date_jan_2017_df 

month        town  flat_type block        street_name storey_range  \
0      2017-01  ANG MO KIO     2 ROOM   406  ANG MO KIO AVE 10     10 TO 12   
1      2017-01  ANG MO KIO     3 ROOM   108   ANG MO KIO AVE 4     01 TO 03   
2      2017-01  ANG MO KIO     3 ROOM   602   ANG MO KIO AVE 5     01 TO 03   
3      2017-01  ANG MO KIO     3 ROOM   465  ANG MO KIO AVE 10     04 TO 06   
4      2017-01  ANG MO KIO     3 ROOM   601   ANG MO KIO AVE 5     01 TO 03   
...        ...         ...        ...   ...                ...          ...   
80369  2020-09      YISHUN     5 ROOM   716       YISHUN ST 71     07 TO 09   
80370  2020-09      YISHUN     5 ROOM   760       YISHUN ST 72     07 TO 09   
80371  2020-09      YISHUN     5 ROOM   835       YISHUN ST 81     04 TO 06   
80372  2020-09      YISHUN  EXECUTIVE   791       YISHUN AVE 2     04 TO 06   
80373  2020-09      YISHUN  EXECUTIVE   387     YISHUN RING RD     04 TO 06   

       floor_area_sqm      flat_model  lease_commence_date  \
0                44.0        Improved                 1979   
1                67.0  New Generation                 1978   
2                67.0  New Generation                 1980   
3                68.0  New Generation                 1980   
4                67.0  New Generation                 1980   
...               ...             ...                  ...   
80369           131.0        Improved                 1987   
80370           122.0        Improved                 1987   
80371           122.0        Improved                 1987   
80372           146.0      Maisonette                 1987   
80373           146.0      Maisonette                 1988   

          remaining_lease  resale_price  
0      61 years 04 months      232000.0  
1      60 years 07 months      250000.0  
2      62 years 05 months      262000.0  
3       62 years 01 month      265000.0  
4      62 years 05 months      265000.0  
...                   ...           ...  
80369  66 years 03 months      440000.0  
80370  65 years 06 months      458000.0  
80371  66 years 04 months      490000.0  
80372  66 years 03 months      558000.0  
80373  66 years 09 months      555000.0  

[80374 rows x 11 columns]

In [67]:
# Check the shape of the dataframe
print(registration_date_jan_2017_df.shape)
print('---------------------------------------')
# Check the datatype information of the features
print(registration_date_jan_2017_df.info())
print('---------------------------------------')
# Check for missing values in each column
print(registration_date_jan_2017_df.isnull().sum())
print('---------------------------------------')
# check for missing values in each column as a percentage of whole dataset
print((registration_date_jan_2017_df.isnull().sum() / registration_date_jan_2017_df .shape[0]) * 100)
print('---------------------------------------')
# Check for duplicates in the dataframe
print(registration_date_jan_2017_df.duplicated().sum())
print('---------------------------------------')

(80374, 11)
---------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 80374 entries, 0 to 80373
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   month                80374 non-null  object 
 1   town                 80374 non-null  object 
 2   flat_type            80374 non-null  object 
 3   block                80374 non-null  object 
 4   street_name          80374 non-null  object 
 5   storey_range         80374 non-null  object 
 6   floor_area_sqm       80374 non-null  float64
 7   flat_model           80374 non-null  object 
 8   lease_commence_date  80374 non-null  int64  
 9   remaining_lease      80374 non-null  object 
 10  resale_price         80374 non-null  float64
dtypes: float64(2), int64(1), object(8)
memory usage: 6.7+ MB
None
---------------------------------------
month                  0
town                   0
flat_type              0
block  

In [68]:
registration_date_jan_2017_df_duplicates = registration_date_jan_2017_df[registration_date_jan_2017_df.duplicated(keep=False)]
print("Registration date Jan 2017 duplicate rows:")
registration_date_jan_2017_df_duplicates.head(20)

Registration date Jan 2017 duplicate rows:


month          town flat_type block       street_name storey_range  \
224   2017-01   BUKIT MERAH    4 ROOM   106    HENDERSON CRES     07 TO 09   
243   2017-01   BUKIT MERAH    4 ROOM   106    HENDERSON CRES     07 TO 09   
304   2017-01  CENTRAL AREA    3 ROOM   271          QUEEN ST     16 TO 18   
305   2017-01  CENTRAL AREA    3 ROOM   271          QUEEN ST     16 TO 18   
505   2017-01   JURONG EAST    4 ROOM   265       TOH GUAN RD     04 TO 06   
510   2017-01   JURONG EAST    4 ROOM   265       TOH GUAN RD     04 TO 06   
590   2017-01   JURONG WEST    4 ROOM  337A      TAH CHING RD     19 TO 21   
591   2017-01   JURONG WEST    4 ROOM  337A      TAH CHING RD     19 TO 21   
671   2017-01     PASIR RIS    4 ROOM   753   PASIR RIS ST 71     01 TO 03   
672   2017-01     PASIR RIS    4 ROOM   753   PASIR RIS ST 71     01 TO 03   
724   2017-01       PUNGGOL    4 ROOM  614B  EDGEFIELD PLAINS     16 TO 18   
725   2017-01       PUNGGOL    4 ROOM  601C      PUNGGOL CTRL     13 TO 15   
736   2017-01       PUNGGOL    4 ROOM  614B  EDGEFIELD PLAINS     16 TO 18   
737   2017-01       PUNGGOL    4 ROOM  601C      PUNGGOL CTRL     13 TO 15   
751   2017-01       PUNGGOL    5 ROOM  171C   EDGEDALE PLAINS     16 TO 18   
764   2017-01       PUNGGOL    5 ROOM  171C   EDGEDALE PLAINS     16 TO 18   
830   2017-01      SENGKANG    4 ROOM  183D    RIVERVALE CRES     04 TO 06   
831   2017-01      SENGKANG    4 ROOM  183D    RIVERVALE CRES     04 TO 06   
1275  2017-02         BEDOK    4 ROOM   156   BEDOK STH AVE 3     10 TO 12   
1283  2017-02         BEDOK    4 ROOM   156   BEDOK STH AVE 3     10 TO 12   

      floor_area_sqm  flat_model  lease_commence_date     remaining_lease  \
224             81.0    Improved                 1975            57 years   
243             81.0    Improved                 1975            57 years   
304             68.0    Improved                 1979  61 years 02 months   
305             68.0    Improved                 1979  61 years 02 months   
505            101.0     Model A                 1998  80 years 09 months   
510            101.0     Model A                 1998  80 years 09 months   
590             92.0     Model A                 2010  92 years 03 months   
591             92.0     Model A                 2010  92 years 03 months   
671            105.0     Model A                 1996  78 years 10 months   
672            105.0     Model A                 1996  78 years 10 months   
724             97.0     Model A                 2012  94 years 11 months   
725             94.0     Model A                 2012  94 years 07 months   
736             97.0     Model A                 2012  94 years 11 months   
737             94.0     Model A                 2012  94 years 07 months   
751            110.0    Improved                 2004  86 years 02 months   
764            110.0    Improved                 2004  86 years 02 months   
830             85.0    Model A2                 2003  85 years 05 months   
831             85.0    Model A2                 2003  85 years 05 months   
1275            84.0  Simplified                 1985  67 years 10 months   
1283            84.0  Simplified                 1985  67 years 10 months   

      resale_price  
224       470000.0  
243       470000.0  
304       470000.0  
305       470000.0  
505       470000.0  
510       470000.0  
590       470000.0  
591       470000.0  
671       368000.0  
672       368000.0  
724       468000.0  
725       470000.0  
736       468000.0  
737       470000.0  
751       468000.0  
764       468000.0  
830       345000.0  
831       345000.0  
1275      470000.0  
1283      470000.0

### Registration_date_jan_2017_df structural observation
- There are 11 columns and 80,374 rows 
- There are 3 datatypes : int64, float64 and object
- There are 210 duplicated rows 
    - The chances of selling the house at the same price are rare. Hence, we will assume it to be duplicated entries.
- Approval_date_1990_1999_df resale_price is int64 while Approval_date_2000_feb_2012_df is float64,
Registration_date_mar_2012_dec_2014_df is float64,
registration_date_jan_2015_dec_2016_df is float64,
Registration_date_jan_2017_df is float64 
- From registration_date_jan_2015_dec_2016_df and Registration_date_jan_2017_df there is an extra column callled 'remaining_lease'
- Registration_date_jan_2017_df 'remaining_lease' is object datatype

TODO: 
1) approval_date_1990_1999_df 
- Remove Duplicates 
- Standarise 'resale_price' datatypes
- new column 'remaining_lease' with 'int64' datatype

2) approval_date_2000_feb_2012_df
- remove duplicates 
- Standarise 'resale_price' datatypes
- new column 'remaining_lease' with 'int64' datatype

3) registration_date_mar_2012_dec_2014_df
- remove duplicates 
- Standarise 'resale_price' datatypes
- new column 'remaining_lease' with 'int64' datatype

4) registration_date_jan_2015_dec_2016_df
- remove duplicates 
- Standarise 'resale_price' datatypes
- new column 'remaining_lease' with 'int64' datatype

5) registration_date_jan_2017_df 
- remove duplicates 
- Standarise 'resale_price' datatypes
- new column 'remaining_lease' with 'int64' datatype


